blockchain/state_manager/src/lib.rs

```diff
@@ -122,6 +134,81 @@ where
         Ok((state_root, rect_root))
     }
 
+    pub async fn tipset_state(&self, tipset: &Tipset) -> Result<(Cid, Cid), Box<dyn StdError>> {
+        span!("tipset_state", {
+            trace!("tipset {:?}", tipset.cids());
+            // if exists in cache return
+            if let Some(cid_pair) = self.cache.read().await.get(&tipset.key()) {
+                return Ok(cid_pair.clone());
+            }
+
+            if tipset.len() == 0 {
+                // NB: This is here because the process that executes blocks requires that the
+                // block miner reference a valid miner in the state tree. Unless we create some
+                // magical genesis miner, this won't work properly, so we short circuit here
+                // This avoids the question of 'who gets paid the genesis block reward'
+                let message_receipts = tipset
+                    .blocks()
+                    .first()
+                    .ok_or_else(|| Error::Other("Could not get message receipts".to_string()))?;
+                let cid_pair = (
+                    tipset.parent_state().clone(),
+                    message_receipts.message_receipts().clone(),
+                );
+                self.cache
+                    .write()
+                    .await
+                    .insert(tipset.key().clone(), cid_pair.clone());
+                return Ok(cid_pair);
+            }
+
+            let block_headers = tipset.blocks();
+            // generic constants are not implemented yet this is a lowcost method for now
+            let cid_pair = self.compute_tipset_state(&block_headers)?;
+            self.cache
+                .write()
+                .await
+                .insert(tipset.key().clone(), cid_pair.clone());
+            Ok(cid_pair)
+        })
+    }
+
+    pub fn compute_tipset_state<'a>(
+        &'a self,
+        blocks: &[BlockHeader],
+    ) -> Result<(Cid, Cid), Box<dyn StdError>> {
+        span!("compute_tipset_state", {
+            if blocks.len() > 2
+                && blocks
+                    .iter()
+                    .zip(blocks.iter().skip(0))
+                    .any(|(a, b)| a.miner_address() == b.miner_address())
```

Comment:
this is just comparing each index with the i+1 ? you need to make sure there are no duplicates, not just no adjacent equivalents